### ISSUE TITLE
Extend CSS tree shaking to account for more scenarios

### ISSUE THREAD

**Issue by asadkn** [ampproject/amp-wp#1492]
To build upon existing style sanitizer, I have to maintain a fork due to lack of hooks. These hooks would be especially useful for the tree shaker. 

- `amp_selector_should_include` would be a great hook for $should_include to allow or remove classes in the tree shaking process. Something like:

`$should_include = apply_filters( 'amp_selector_should_include', $should_include, $selector, $parsed_selector );`

I am using this to write a custom check to strip out few more classes that linger around despite being unnecessary, like some multi-class selectors such as `.something.else` remaining even when unused. This filter can essentially whitelist classes that might be wrongly removed, or blacklist classes that are unnecessarily preserved. 

- Or more importantly, allowing filtering of CSS parts would be great. Just before `$sheet_size` is calculated, a filter such as following could be great: 

`$stylesheet = apply_filters( 'amp_stylesheet_part', $stylesheet );`

Using this filter, a plugin/theme can do further processing to remove CSS as necessary. I am using it to filter empty `@media` and `@support` blocks but it could also be used for addition or further removal. 

P.S. thanks for the tremendous effort on v1.0. I have been using the beta in production and have no complaints so far.

**Comment by westonruter** [comment]
Adding filters _inside_ of a sanitizer is not going to be viable because this will break caching. Any filtering needs to be done at the level of the args being passed into the sanitizers, such as via the `amp_content_sanitizers` filter, which is improved in #1478. 

It sounds like the more fundamental issue is that the tree shaking should be extended to handle more cases. Alternatively, you should consider just not enqueueing the CSS to begin with if you know it is not going to be used on a page. That is the very best.

**Comment by asadkn** [comment]
As long as the devs are aware of the cache at this filter, I don't think that should prevent addition of filters or hooks. Filters at a _processing_ level are perfectly fine. 

While the proponents of AMP would prefer "native" AMP everywhere, the reality is "paired" is the practical mode that most are adopting. Tree shaking has been the best addition to this plugin and having the ability to extend and cover edge cases makes life easy for developers. There are always going to be cases the tree shaker will need to be extended for. 

For example, advanced processing to handle all cases of `.carousel.dark` is very complex and expensive and when there are multiple versions of `.dark` in CSS. Devs can manually handle the case to keep `.carousel.dark` but remove `.slider.dark`. There are some cases like these which shouldn't be handled in the core due to processing overhead but a dev should have a choice to extend.

As for my other example previously of stripping empty `@media` and `@support` groups, that _should_ be handled by core but it still gives devs the freedom to do post-processing for edge cases.

**Comment by westonruter** [comment]
> As long as the devs are aware of the cache at this filter, I don't think that should prevent addition of filters or hooks. Filters at a _processing_ level are perfectly fine.

Filters at the processing level are the ones that are most problematic for caching. The sanitizers should be considered a black box that take an input and an output. If the sanitizers have previously computed the output for a given input, then they can short-circuit be serving the cached output for the given input.

This doesn't have anything to do with native vs paired. The post processor (sanitizer) caching applies for both.

**Comment by asadkn** [comment]
> This doesn't have anything to do with native vs paired. The post processor (sanitizer) caching applies for both.

I wasn't implying it has anything to do with native vs paired. To clarify, I was emphasizing not everything is built for native AMP and this is why the paired mode - AMP being an afterthought - is the reality. Hence, edge-cases do exist that can't be covered by the tree shaker and extending it usually is the sensible course of action.

I know about the cache implications here and any extension can be aware of the same. While I don't agree, I understand your preference.

Alternatively, can the style sanitizer be made _open for extension_? Right now almost every method is private. It would help if they were protected.

**Comment by westonruter** [comment]
Thanks. One reason for making the methods private for the sanitizers is that it reduces the API surface area that the plugin has to maintain. In the future I expect more of the methods can be made public, but they are private for 1.0. 

It would be helpful to see some examples of how you intend the `amp_selector_should_include` filter to work. Maybe there is an alternative. However, is this not an option:

> It sounds like the more fundamental issue is that the tree shaking should be extended to handle more cases. Alternatively, you should consider just not enqueueing the CSS to begin with if you know it is not going to be used on a page. That is the very best.

?

**Comment by asadkn** [comment]
> It would be helpful to see some examples of how you intend the amp_selector_should_include filter to work.

There are skinning systems used based on multiple compounded classes in the selector, for instance:

```css
.block {}
.block.dark {}
.block-b {} 
.block-b.dark {}
.block-news {}
.block-news.dark {}
```

If `class="block dark"` exists on the page, even if rest of blocks are just `class="block-b"` and so on, their .dark counterparts and all the shared child classes will be preserved. Removing them saves about 16KB of CSS in this case.

Don't have the exact code with me right now, but I am using a custom sanitizer to first collect (for performance) all the block classes present in the DOM using for ex: `$this->xpath->query('//div[ contains(@class, "block") ]')` (for brevity). Saved in a `$classes_present_in_dom` array.

`amp_selector_should_include` used something like this:

```php
	public function selector_should_include($should_include, $selector, $parsed) 
	{
		// This is getting removed already
		if (!$should_include) {
			return false;
		}
		
		// Joined classes?
		if (preg_match('/\.block([a-zA-Z0-9_\-]*)\.([a-zA-Z0-9_\-\.]+)/', $selector, $match)) {
			$sel_classes   = explode('.', $match[1]);
			$block_classes = $this->classes_present_in_dom[ $sel_classes[0] ];

			// Not all classes in DOM? Don't include this selector
			if (array_diff($sel_classes, $block_classes)) {
				return false;
			}
		}

		return true;
	}
```

A generic solution for handling multiple classes can be as good as you already have - anything more specific would be too resource intensive. I started with a generic solution first to create a PR, but  realized the performance hit.

**Comment by westonruter** [comment]
The tree shaker currently obtains a list of all classes used in the doc. It seems like there could be an efficient way to get all class names used together and then find the intersection between these and the concatenated class names used in style rules. We'd need to carefully optimize the data structure and algorithm to not adversely impact performance.

**Comment by westonruter** [comment]
In addition to shaking compound class selectors such as `.block-news.dark`, other key selectors to shake include (from Twenty Nineteen):

* Different class names used in descendant selectors:`.entry .entry-header`
* The same class name used in descendant selectors: `.sub-menu .sub-menu .sub-menu .sub-menu`
* Child and sibling selectors: `.main-navigation .main-menu > li > a + svg`
* Element selectors used in descendant selectors: `.widget_categories ul ul ul ul ul li > a`

**Comment by westonruter** [comment]
Here is CSS with Twenty Nineteen that escaping the tree-shaker: https://gist.github.com/westonruter/38e465dfcb43d6692857da6c8ce68d3b

**Comment by westonruter** [comment]
Just noticed this incredibly meta selector: `.wp-block-quote[style*="text-align:right"]`. That seems like a hack. It will never match elements in the AMP plugin as it stands today, since all `style` elements are removed.

**Comment by westonruter** [comment]
One thing which can be done for Twenty Nineteen is eliminate the selectors that mention `[focus-within]` since those polyfill rules will not work in AMP (since they require JS). That eliminates 1.5KB of CSS.

**Comment by westonruter** [comment]
As a next step for improving the tree shaker further, it seems feasible to tree-shake selectors that contain multiple class names on a given element in a selector. For example:

```html
<style type="text/css">
span.bg-dark.fg-light {
    background-color: black;
    color: white;
}
</style> 

<span>hello!</span>
<div class="bg-dark">dark background</div>
<div class="fg-light">light foreground</div>
```

The CSS selector currently is not tree-shaken because the document contains a `span` and classes `bg-dark` and `fg-light`. In reality, it the selector should be removed because there is no `span` that has both the classes `bg-dark` and `fg-light`.

To do this, we'll need to change how we gather the tag names, IDs, attributes, class names selectors to organize them by the elements they are connected to inside of a selector, not just that they are mentioned _anywhere_ in a selector. This would benefit from #2102 to have more robust CSS selector parser.

**Comment by westonruter** [comment]
Note that this will be less important to do now that AMP is poised to bump the limit from 50K to 75K. See https://github.com/ampproject/amphtml/issues/26466.

Nevertheless, if we did add deeper tree shaking…

A standard part of a good scalable WordPress site is to have a page caching plugin installed. This is used to serve responses to users who are not logged-in. So in practice normal site visitors should not have to wait for the CSS minification process to complete before seeing a page. 

The AMP plugin is not currently providing proactive guidance to install such a page caching plugin, but we have this on our radar. If we could reliably detect when page caching is enabled, then this could mean that we perform extra expensive tree shaking operations when the response is being served to an unauthenticated user which would then be cached. For other users who are logged-in, we could go ahead and serve less-minified CSS but slap on the `data-ampdevmode` attribute to `style[amp-custom]` and exceed the 50KB limit. For such authenticated users, we are already serving AMP pages in dev mode for the sake of the WordPress admin bar.

The user's authentication state is something I've pondered as a signal for whether to [minify CSS class names](https://github.com/ampproject/amp-wp/issues/1518#issuecomment-571249081). Nevertheless, minifying class names introduces new opportunities for CSS to break especially when taking into account dynamic documents (e.g. `<amp-live-list>`) so I want to make this a last resort for minification. (This also makes debugging very difficult.)

So all this to say: the AMP plugin could invest in more sophisticated CSS tree shaking as long as we have in place detection that page caching is enabled.
